File: sdks/php/src/Api/SendingDomains.php
<?php

declare(strict_types=1);

namespace Apollo\Deploy\Signal\Sdk\Api;

use Apollo\Deploy\Signal\Sdk\SDKError;
use Apollo\Deploy\Signal\Sdk\SSEEvent;
use Apollo\Deploy\Signal\Sdk\Transport;
use Apollo\Deploy\Signal\Sdk\DomainListPageResponse;
use Apollo\Deploy\Signal\Sdk\DomainResponse;
use Apollo\Deploy\Signal\Sdk\RegisterDomainRequest;
use Apollo\Deploy\Signal\Sdk\UpdateBimiRequest;
use Psr\Http\Message\ResponseInterface;

/** SendingDomainsAPI domain operations. */
class SendingDomains
{
    public function __construct(private Transport $transport)
    {
    }

    /**
     * @param string $projectId
     * @return DomainListPageResponse
     * @throws SDKError
     */
    public function listDomains(
        string $projectId,
    ): DomainListPageResponse
    {
        $path = '/v1/projects/${projectId}/domains';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(DomainListPageResponse::class, 'fromArray')) {
            return DomainListPageResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $domainId
     * @return DomainResponse
     * @throws SDKError
     */
    public function getDomain(
        string $projectId,
        string $domainId,
    ): DomainResponse
    {
        $path = '/v1/projects/${projectId}/domains/${domainId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${domainId}', rawurlencode(Transport::wireValue($domainId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(DomainResponse::class, 'fromArray')) {
            return DomainResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param RegisterDomainRequest $input
     * @return DomainResponse
     * @throws SDKError
     */
    public function registerDomain(
        string $projectId,
        RegisterDomainRequest $input,
    ): DomainResponse
    {
        $path = '/v1/projects/${projectId}/domains';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(DomainResponse::class, 'fromArray')) {
            return DomainResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $domainId
     * @return DomainResponse
     * @throws SDKError
     */
    public function verifyDomain(
        string $projectId,
        string $domainId,
    ): DomainResponse
    {
        $path = '/v1/projects/${projectId}/domains/${domainId}/verify';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${domainId}', rawurlencode(Transport::wireValue($domainId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(DomainResponse::class, 'fromArray')) {
            return DomainResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $domainId
     * @return void
     * @throws SDKError
     */
    public function deleteDomain(
        string $projectId,
        string $domainId,
    ): void
    {
        $path = '/v1/projects/${projectId}/domains/${domainId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${domainId}', rawurlencode(Transport::wireValue($domainId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'DELETE',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $this->transport->readResponseBody($response);
        return;
    }

    /**
     * @param string $projectId
     * @param string $domainId
     * @return DomainResponse
     * @throws SDKError
     */
    public function verifyBimi(
        string $projectId,
        string $domainId,
    ): DomainResponse
    {
        $path = '/v1/projects/${projectId}/domains/${domainId}/bimi/verify';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${domainId}', rawurlencode(Transport::wireValue($domainId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(DomainResponse::class, 'fromArray')) {
            return DomainResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $domainId
     * @param UpdateBimiRequest $input
     * @return DomainResponse
     * @throws SDKError
     */
    public function updateBimi(
        string $projectId,
        string $domainId,
        UpdateBimiRequest $input,
    ): DomainResponse
    {
        $path = '/v1/projects/${projectId}/domains/${domainId}/bimi';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${domainId}', rawurlencode(Transport::wireValue($domainId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'PATCH',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(DomainResponse::class, 'fromArray')) {
            return DomainResponse::fromArray($data);
        }
        return $data;
    }

    private function decodeResponse(ResponseInterface $response): mixed
    {
        $contents = $this->transport->readResponseBody($response);
        if (trim($contents) === '') {
            $requestId = $response->getHeaderLine('x-request-id')
                ?: $response->getHeaderLine('request-id')
                ?: null;
            throw new SDKError(
                message: 'Expected a response body for a non-void operation',
                status: $response->getStatusCode(),
                code: 'RESPONSE_BODY_MISSING',
                requestId: $requestId,
            );
        }

        try {
            return json_decode($contents, true, 512, JSON_THROW_ON_ERROR);
        } catch (\JsonException) {
            return $contents;
        }
    }
}

File: sdks/rust/src/api/sending_domains.rs
use crate::error::SdkError;
use crate::transport::Transport;
use reqwest::Method;
use serde::Serialize;
use crate::types::{ DomainListPageResponse, DomainResponse, RegisterDomainRequest, UpdateBimiRequest };

/// SendingDomainsAPI API operations.
pub struct SendingDomainsAPI {
    transport: Transport,
}

impl SendingDomainsAPI {
    pub(crate) fn new(transport: Transport) -> Self {
        Self { transport }
    }

    pub async fn list_domains(
        &self,
        project_id: String,
    ) -> Result<DomainListPageResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/domains".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<DomainListPageResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn get_domain(
        &self,
        project_id: String,
        domain_id: String,
    ) -> Result<DomainResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/domains/${domainId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${domainId}", &encode_path(&domain_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<DomainResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn register_domain(
        &self,
        project_id: String,
        body: &RegisterDomainRequest,
    ) -> Result<DomainResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/domains".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<DomainResponse>(
            Method::POST,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await
    }

    pub async fn verify_domain(
        &self,
        project_id: String,
        domain_id: String,
    ) -> Result<DomainResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/domains/${domainId}/verify".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${domainId}", &encode_path(&domain_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<DomainResponse>(
            Method::POST,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn delete_domain(
        &self,
        project_id: String,
        domain_id: String,
    ) -> Result<(), SdkError> {
        let mut path = "/v1/projects/${projectId}/domains/${domainId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${domainId}", &encode_path(&domain_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute_raw(
            Method::DELETE,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await.map(|_| ())
    }

    pub async fn verify_bimi(
        &self,
        project_id: String,
        domain_id: String,
    ) -> Result<DomainResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/domains/${domainId}/bimi/verify".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${domainId}", &encode_path(&domain_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<DomainResponse>(
            Method::POST,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn update_bimi(
        &self,
        project_id: String,
        domain_id: String,
        body: &UpdateBimiRequest,
    ) -> Result<DomainResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/domains/${domainId}/bimi".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${domainId}", &encode_path(&domain_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<DomainResponse>(
            Method::PATCH,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await
    }

}

fn encode_path(value: &str) -> String {
    urlencoding::encode(value).into_owned()
}

fn encode_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    match serde_json::to_value(value)
        .map_err(|error| SdkError::ValidationError(error.to_string()))?
    {
        serde_json::Value::String(value) => Ok(value),
        value => Ok(value.to_string()),
    }
}

fn encode_cookie_name(name: &str) -> String {
    urlencoding::encode(name).into_owned()
}

fn encode_cookie_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    let value = encode_value(value)?;
    Ok(urlencoding::encode(&value).into_owned())
}
